Fixes, permissions and versioning

gnoxr opened the pull request:
What does this pull request contain?


**Server-side fixes**
- Refactoring of some files; there is now a single `utils.js` file
- `unfoldIds()` is now calles by the routes instead of the DB
- Response arrays are streamed as fast as possible
- `upload` and `version` has been updated accordingly to the new DB specifications. **Update**: read newer comments about `upload` and `version`.

**Server-side novelties**
- Fix #94: `graph` is now possible with a `depth` option. It does not impact existing code.
- There are sample asynchronous `post` event listeners.
- New tests for `version`
- Fix #66: The permission system has been implemented (`guest`, `user`, `editor`, `admin`) :smiley: 

**Client-side**
- JS: return `null` instead of `false` on errors (except the functions returning booleans)
- Python: drop the URL encoding for the `search` and `search_one` methods, as it works better.


I will add new changes to this pull request whenever possible.

remyla commented on server-nodejs/index.js:
```diff
@@ -39,28 +33,33 @@ if (module.parent) {
+    var port = process.env.HTTP_PORT || 8090;
+    debug('Creating HTTP server on port %s', port);
+    global.server = require('http').createServer(app);
+    startServer('HTTPS', port);
```
should be http according to what you want to do

gnoxr commented:
Important update: the `version` *and* `update` routes work!

When we upload a file, it will:
- If the request body contains an id:
  - Save any existing file with the same name *if its node exists in the database*
  - Move the new file at the specified path
  - Create a new `version()`
- Otherwise:
  - Move the new file at the specified path
  - `create()` a new version

The name is made of parts, separated by `/`. The `time` is inserted before the first point of the last part of the name, or at the end of this part if there is no `.`, and all of this in a single RegExp by @zankia.
```javascript
"/path/explosion.tar.gz"  =>  "/path/explosion_TIME.tar.gz"
"/path/explosion.jpg/"    =>  "/path/explosion_TIME.jpg/"
"/path/explosion/"        =>  "/path/explosion_TIME/"
""                        =>  "_TIME"
```

The `version()` method will now:
- Create a copy of the existing node into a new one (with a new id), changing:
  - The `file` name, replaced by the version name (with the `time`)
  - The `#parent` key, referring to its previous id (the head id)
  - The `version` key, removed
- Update the original node:
  - Delete the previous keys
  - Set the new keys (at least `time`, `file` and `author`, plus the ones specified in the request body)
  - Set the `version` to 1 + count of nodes referring to it as their `#parent`

remyla commented:
Thank you very much for all this work.

I would like to have a good understanding about compatibility breaking parts, especially for the Zombillenium production which is tightly connected to the engine.

* /api/version is breaking compatibility
* we need to add a `class` key to current user nodes

if this is only a matter of keeping /api/version for the current production, we could add the new operation under a new name

about exposing routes handlers in all routes, why not, I understand that you were searching for a code factorization closer to what we had in the Php implementation :) why not.

it is very interesting to implement this code in running productions, then what are the other points that would break the compatibility? it may be interesting to separate them.

Merci!

gnoxr commented:
AFAIK (I just finished to read the diff) the things breaking compatibility are:
- `update` and `version`, which now save the existing HEAD if there was one
- The JS client returning `null` instead of `false` for most routes (I understood that it was not an issue as the JS client is not being currently used)
- The control properties can be overwritten in the `create` method (useful for the other routes and maybe for imports as well + anyone could `update` them anyways)
- Yes, we need to add a `class` to existing users if we dont want them to be treated as `guest`s. Or maybe we could set the default class directly to `user` or `editor` in `auth-jwt-node`, as the users will be rejected anyways if their node doesn't exist in the DB.

I updated the two `auth` middlewares so that they automagically set a `class` key to the current user whenever there is none. Also note that there is no permission required for the `upload` route. This can be changed.

Things not breaking compatibility but really looking like they do:
- Removing urlencoding in Python searches (it was breaking compatibility as we had to type `\/` instead of `/`)
- The `graph` depth, set to 0 by default, and the library specifications were not changed (server-only update for now)

I will put back the previous `version` and rename the current one, and perhaps set a permission level for the upload if you tell me which class has this permission. User or editor?

remyla commented:
here is my feedback about permissions
* the authentication only needs a username and password as input. if we decide to prepare some data in it, we quit the "do one thing and do it right" paradigm for this file.
* to keep the permissions system generic it is more convenient not to "hardcode" values for classes. We only need these values: - on the user nodes in the db, - in the configuration file for permissions to perform the comparisons
* the default level for the permissions if an authentication is available is `guest`
* the upload is available for every users above `user` class (not `guest`)

remyla commented:
About the operation based permissions:
* having a separate configuration file like `permissions.json` is nice because if the authentication is set to "no" (which is the default) there is no need for it
* the wildcard `*` is handy, mainly to accept every read operations for every classes, and for any added class. Any added class on nodes will not be able to use these operations otherwise, in a straightforward manner without having to modify the configuration file

* adding further modifications to these pull request makes it more difficult for me to merge it as some of its commits break compatibility. I will be cherry picking each commit otherwise. I think that the operation-based permissions don't break the compatibility therefore it could be fastly merged if well seperated.

thank you!

remyla commented:
From your commits the 10th June, cherry picked:

 Refactorization
@gnoxr gnoxr committed on Jun 10 
commit 658e13432d38d4904bce590ace93945b6cf2f984

Author: gnoxr <quester@protonmail.com>
Date:   Fri Jun 10 18:38:43 2016 +0200

    Proper way to require the client-side modules

commit 1ea26bb6b6e5ce536714c02cfc906304ec450cf4
Author: gnoxr <quester@protonmail.com>
Date:   Fri Jun 10 18:26:43 2016 +0200

    Return null in the JS client

commit d48f056932058b4279439c76d6842692c55f4ec3
Author: gnoxr <quester@protonmail.com>
Date:   Fri Jun 10 17:55:40 2016 +0200

    Call unfoldIds() from the routes
